python/pylibvw.cc

```diff
@@ -542,32 +542,85 @@ std::string varray_char_to_string(v_array<char>& a)
 
 std::string my_get_tag(example_ptr ec) { return varray_char_to_string(ec->tag); }
 
-uint32_t ex_num_namespaces(example_ptr ec) { return (uint32_t)ec->indices.size(); }
+uint32_t ex_num_namespaces(example_ptr ec) { return (uint32_t)ec->feature_space.size(); }
 
-unsigned char ex_namespace(example_ptr ec, uint32_t ns) { return ec->indices[ns]; }
+unsigned char ex_namespace(example_ptr ec, uint32_t distance) 
+{
+  auto it = ec->begin();
+  for (int i = 0; i < distance; ++i, ++it);
+  return it.index();
+}
 
-uint32_t ex_num_features(example_ptr ec, unsigned char ns) { return (uint32_t)ec->feature_space[ns].size(); }
+uint32_t ex_num_features(example_ptr ec, char* ns) {
+  auto fs = ec->feature_space;
+  uint64_t hash = hashstring(ns, strlen(ns), 0);
+  return (strlen(ns) == 1) ? fs.namespace_index_end_proxy(ns[0]) - fs.namespace_index_begin_proxy(ns[0]) : fs[hash].size();
+}
 
-uint32_t ex_feature(example_ptr ec, unsigned char ns, uint32_t i)
+uint32_t ex_feature(example_ptr ec, char* ns, uint32_t i)
 {
-  return (uint32_t)ec->feature_space[ns].indicies[i];
+  uint64_t hash = hashstring(ns, strlen(ns), 0);
+  if (strlen(ns) == 1)
+  {
+    auto it = ec->feature_space.namespace_index_begin_proxy(ns[0]);
+    it += i;
+    return (*it).index();
+  }
+  else
+  {
+    auto it = (ec->feature_space[hash]).begin();
+    it += i;
+    return (*it).index();
+  }
 }
 
-float ex_feature_weight(example_ptr ec, unsigned char ns, uint32_t i) { return ec->feature_space[ns].values[i]; }
+float ex_feature_weight(example_ptr ec, char* ns, uint32_t i)
+{ 
+  uint64_t hash = hashstring(ns, strlen(ns), 0);
+  if (strlen(ns) == 1)
+  {
+    auto it = ec->feature_space.namespace_index_begin_proxy(ns[0]);
+    it += i;
+    return (*it).value();
+  }
+  else
+  {
+    auto it = (ec->feature_space[hash]).begin();
+    it += i;
+    return (*it).value();
+  }
+}
 
-float ex_sum_feat_sq(example_ptr ec, unsigned char ns) { return ec->feature_space[ns].sum_feat_sq; }
+float ex_sum_feat_sq(example_ptr ec, char* ns)
+{
+  auto fs = ec->feature_space;
+  if (strlen(ns) == 1) 
+  {
+    auto beg = fs.namespace_index_begin(ns[0]);
+    auto end = fs.namespace_index_end(ns[0]);
+    float sum_feat_sq = 0.f;
+    for (auto it = beg; it != end; ++it) sum_feat_sq += (*it).sum_feat_sq;
```

Comment:
You can use `namespace_index_range` for cases like this

Reply:
This works for most cases, however if you need to access the hash I dont think this is possible